torch/csrc/distributed/c10d/comm.cpp

```diff
@@ -79,4 +80,29 @@ void broadcast_coalesced(
   }
 }
 
+GradBucket::GradBucket(std::vector<at::Tensor>& tensors) : tensors_(tensors){};
+
+std::vector<at::Tensor> GradBucket::get_tensors() {
+  return tensors_;
+};
+
+PythonCommHook::PythonCommHook(py::object state, py::object hook)
+    : state_(std::move(state)), hook_(std::move(hook)){};
+c10::intrusive_ptr<torch::jit::Future> PythonCommHook::operate(
+    const GradBucket& bucket) {
+  py::gil_scoped_acquire acquire;
+
+  c10::intrusive_ptr<torch::jit::Future> fut;
+  return hook_(state_, bucket)
+      .cast<std::shared_ptr<torch::jit::PythonFutureWrapper>>()
```

Comment:
what if cast throws an error?

Reply:
Included a try catch, and an additional condition in `test_ddp_invalid_comm_hook`. In that unit test we are checking 4 things. I thing it is not a problem, but let me know if you want to divide the test into pieces.